File: src/main/java/org/padler/gradle/minify/minifier/JsMinifier.java
package org.padler.gradle.minify.minifier;

import com.google.common.collect.ImmutableList;
import com.google.javascript.jscomp.*;
import org.padler.gradle.minify.minifier.options.JSMinifierOptions;
import org.padler.gradle.minify.minifier.result.Error;
import org.padler.gradle.minify.minifier.result.Warning;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.Collections;
import java.util.List;

public class JsMinifier extends Minifier {

    protected JSMinifierOptions minifierOptions = new JSMinifierOptions();

    private CompilerOptions options = new CompilerOptions();

    @Override
    @SuppressWarnings("java:S2259") // suppresses false positive at sourcemapFile.getName()
    protected void minifyFile(File srcFile, File dstFile) throws IOException {
        com.google.javascript.jscomp.Compiler compiler = new com.google.javascript.jscomp.Compiler();
        setOptions();

        List<SourceFile> externs = AbstractCommandLineRunner.getBuiltinExterns(new CompilerOptions().getEnvironment());
        SourceFile sourceFile = SourceFile.fromFile(srcFile.getAbsolutePath());

        File sourcemapFile = null;
        if (Boolean.TRUE.equals(minifierOptions.getCreateSoureMaps())) {
            sourcemapFile = new File(dstFile.getAbsolutePath() + ".map");
            options.setSourceMapOutputPath(sourcemapFile.getAbsolutePath());
            options.setSourceMapLocationMappings(Collections.singletonList(new RelativePathLocationMapping()));
        }

        Result result = compiler.compile(externs, ImmutableList.of(sourceFile), options);

        if (result.success) {
            String source = compiler.toSource();

            if (Boolean.TRUE.equals(minifierOptions.getCreateSoureMaps())) {
                StringBuilder sourceMapContent = new StringBuilder();
                result.sourceMap.appendTo(sourceMapContent, dstFile.getName());
                writeToFile(sourcemapFile, sourceMapContent.toString());

                source += "\n//# sourceMappingURL=" + sourcemapFile.getName();
            }
            writeToFile(dstFile, source);
        } else {
            for (JSError error : result.errors) {
                report.add(new Error(error));
            }
            for (JSError warning : result.warnings) {
                report.add(new Warning(warning));
            }
        }

    }

    private void setOptions() {
        minifierOptions.getCompilationLevel().setOptionsForCompilationLevel(options);
        options.setEnvironment(minifierOptions.getEnv());
        if (minifierOptions.getLanguageIn() != null)
            options.setLanguageIn(minifierOptions.getLanguageIn());
        if (minifierOptions.getLanguageOut() != null)
            options.setLanguageOut(minifierOptions.getLanguageOut());

        minifierOptions.getWarningLevel().setOptionsForWarningLevel(options);
        options.setExtraAnnotationNames(minifierOptions.getExtraAnnotationNames());
        options.setStrictModeInput(minifierOptions.getStrictModeInput());

        if (Boolean.TRUE.equals(minifierOptions.getDebug()))
            minifierOptions.getCompilationLevel().setDebugOptionsForCompilationLevel(options);

        options.setExportLocalPropertyDefinitions(minifierOptions.getExportLocalPropertyDefinitions());

        for (CommandLineRunner.FormattingOption formattingOption : minifierOptions.getFormatting()) {
            switch (formattingOption) {
                case PRETTY_PRINT:
                    options.setPrettyPrint(true);
                    break;
                case PRINT_INPUT_DELIMITER:
                    options.printInputDelimiter = true;
                    break;
                case SINGLE_QUOTES:
                    options.setPreferSingleQuotes(true);
                    break;
                default:
                    throw new IllegalStateException("Unknown formatting option: " + this);
            }
        }

        options.setGenerateExports(minifierOptions.getGenerateExports());
        options.setRenamePrefixNamespace(minifierOptions.getRenamePrefixNamespace());
        options.setRenamePrefix(minifierOptions.getRenameVariablePrefix());
        options.setModuleResolutionMode(minifierOptions.getModuleResolution());
        options.setProcessCommonJSModules(minifierOptions.getProcessCommonJsModules());
        options.setPackageJsonEntryNames(minifierOptions.getPackageJsonEntryNames());
        options.setAngularPass(minifierOptions.getAngularPass());
        options.setDartPass(minifierOptions.getDartPass());
        options.setForceLibraryInjection(minifierOptions.getForceInjectLibrary());
        options.setPolymerVersion(minifierOptions.getPolymerVersion());
        options.setRewritePolyfills(minifierOptions.getRewritePolyfills());
        options.setOutputCharset(minifierOptions.getCharset());
        options.setChecksOnly(minifierOptions.getChecksOnly());
        if (minifierOptions.getBrowserFeaturesetYear() != null)
            options.setBrowserFeaturesetYear(minifierOptions.getBrowserFeaturesetYear());
    }

    @Override
    protected boolean fileTypeMatches(Path f) {
        return "js".equals(getExtension(f.toString()));
    }

    @Override
    public JSMinifierOptions getMinifierOptions() {
        return minifierOptions;
    }

    @Override
    protected String getMinifierName() {
        return "JS Minifier";
    }

    @Override
    protected String rename(String oldName) {
        return oldName.replace(".js", ".min.js");
    }
}

File: src/main/kotlin/org/padler/gradle/minify/minifier/JsMinifier.kt
package org.padler.gradle.minify.minifier

import com.google.common.collect.ImmutableList
import com.google.javascript.jscomp.*
import org.padler.gradle.minify.minifier.options.JSMinifierOptions
import org.padler.gradle.minify.minifier.result.Error
import org.padler.gradle.minify.minifier.result.Warning
import java.io.File
import java.nio.file.Path

class JsMinifier(override var minifierOptions: JSMinifierOptions = JSMinifierOptions()) : Minifier() {

    private val options = CompilerOptions()

    override fun minifyFile(srcFile: File, dstFile: File) {
        val compiler = Compiler()
        setOptions()
        val externals = AbstractCommandLineRunner.getBuiltinExterns(CompilerOptions().environment)
        val sourceFile = SourceFile.fromFile(srcFile.absolutePath)
        var sourcemapFile: File? = null
        if (minifierOptions.createSourceMaps) {
            sourcemapFile = File(dstFile.absolutePath + ".map")
            options.setSourceMapOutputPath(sourcemapFile.absolutePath)
            options.setSourceMapLocationMappings(listOf(SourceMap.LocationMapping { location ->
                val index = location.lastIndexOf('/')
                if (index != -1) location.substring(index + 1) else null
            }))
        }
        val result = compiler.compile(externals, ImmutableList.of(sourceFile), options)
        if (result.success) {
            var source = compiler.toSource()
            if (minifierOptions.createSourceMaps) {
                val sourceMapContent = StringBuilder()
                result.sourceMap.appendTo(sourceMapContent, dstFile.name)
                writeToFile(sourcemapFile!!, sourceMapContent.toString())
                source += "\n//# sourceMappingURL=${sourcemapFile.name}"
            }
            writeToFile(dstFile, source!!)
        } else {
            for (error in result.errors) {
                report.add(Error(error))
            }
            for (warning in result.warnings) {
                report.add(Warning(warning))
            }
        }
    }

    private fun setOptions() {
        minifierOptions.compilationLevel.setOptionsForCompilationLevel(options)
        options.environment = minifierOptions.env
        if (minifierOptions.languageIn != null) options.languageIn = minifierOptions.languageIn
        if (minifierOptions.languageOut != null) options.setLanguageOut(minifierOptions.languageOut)
        minifierOptions.warningLevel.setOptionsForWarningLevel(options)
        options.setExtraAnnotationNames(minifierOptions.extraAnnotationNames)
        options.setStrictModeInput(minifierOptions.strictModeInput)
        if (minifierOptions.debug) minifierOptions.compilationLevel
                .setDebugOptionsForCompilationLevel(options)
        options.setExportLocalPropertyDefinitions(minifierOptions.exportLocalPropertyDefinitions)
        for (formattingOption in minifierOptions.formatting) {
            when (formattingOption) {
                CommandLineRunner.FormattingOption.PRETTY_PRINT -> options.isPrettyPrint = true
                CommandLineRunner.FormattingOption.PRINT_INPUT_DELIMITER -> options.printInputDelimiter = true
                CommandLineRunner.FormattingOption.SINGLE_QUOTES -> options.setPreferSingleQuotes(true)
                else                                                     -> throw IllegalStateException("Unknown formatting option: $this")
            }
        }
        options.setGenerateExports(minifierOptions.generateExports)
        options.setRenamePrefixNamespace(minifierOptions.renamePrefixNamespace)
        options.setRenamePrefix(minifierOptions.renameVariablePrefix)
        options.moduleResolutionMode = minifierOptions.moduleResolution
        options.processCommonJSModules = minifierOptions.processCommonJsModules
        options.packageJsonEntryNames = minifierOptions.packageJsonEntryNames
        options.setAngularPass(minifierOptions.angularPass)
        options.setDartPass(minifierOptions.dartPass)
        options.setForceLibraryInjection(minifierOptions.forceInjectLibrary)
        options.setPolymerVersion(minifierOptions.polymerVersion)
        options.rewritePolyfills = minifierOptions.rewritePolyfills
        options.setOutputCharset(minifierOptions.charset)
        options.setChecksOnly(minifierOptions.checksOnly)
        if (minifierOptions.browserFeaturesetYear != null) options.browserFeaturesetYear = minifierOptions
                .browserFeaturesetYear
    }

    override fun fileTypeMatches(f: Path): Boolean {
        return "js" == getExtension(f.toString())
    }

    override val minifierName = "JS Minifier"

    override fun rename(oldName: String): String {
        return oldName.replace(".js", ".min.js")
    }
}
